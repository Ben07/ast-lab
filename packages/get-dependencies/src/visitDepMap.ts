import { DependencyMap, Entry } from "ast-lab-types";
import _debug from 'debug';

const debug = _debug('get-dependencies:visit');

type Visited = { [module: string]: { [member: string]: 1 } };

export default function visitDependencyMap(dependencyMap: DependencyMap, entries: Entry[]): Visited {
  const visited = {} as Visited;
  let entryQueue = entries;
  debug('visit entries:', entries);
  while(entryQueue.length) {
    const { source: mod, name } = entryQueue.shift() as Entry;
    let modVisited = visited[mod];
    if (!modVisited) {
      modVisited = visited[mod] = {};
    }
    const affected = dependencyMap.get(mod);
    debug(`${mod} affects`, affected);
    if (affected && !modVisited[name]) {
      const affectedEntries = affected.get(name);
      debug('current queue:', entryQueue);
      debug('new to queue:', affectedEntries);
      if (affectedEntries && affectedEntries.length) {
        entryQueue = entryQueue.concat(affectedEntries);
      }
    }
  }
  return visited;
}