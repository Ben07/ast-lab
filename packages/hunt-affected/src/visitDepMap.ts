import { DependencyMap, Entry, Member } from "ast-lab-types";
import _debug from 'debug';
import { MODULE_ALL } from "./constants";

const debug = _debug('hunt-affected:visit');

export type Visited = { [module: string]: Set<Member> };

/**
 * Walk through the dependency map from given entries to find out what are affected.
 * @param dependencyMap 
 * @param entries 
 */
export default function visitDependencyMap(dependencyMap: DependencyMap, entries: Entry[]): Visited {
  const visited = {} as Visited;
  let entryQueue = entries;
  debug('visit entries:', entries);
  while(entryQueue.length) {
    const { source: mod, name } = entryQueue.shift() as Entry;
    let modVisited = visited[mod];
    if (!modVisited) {
      modVisited = visited[mod] = new Set();
    }
    const affected = dependencyMap.get(mod);
    debug(`${mod} affects`, affected);
    if (affected && !modVisited.has(name)) {
      modVisited.add(name);
      const matchedEntries = affected.get(name);
      debug('current queue:', entryQueue);
      if (matchedEntries && matchedEntries.length) {
        debug('new to queue:', matchedEntries);
        entryQueue = entryQueue.concat(matchedEntries);
      }
      const allEntries = affected.get(MODULE_ALL);
      if (allEntries &&  allEntries.length) {
        debug('new to queue:', allEntries);
        entryQueue = entryQueue.concat(allEntries);
      }
    }
  }
  return visited;
}