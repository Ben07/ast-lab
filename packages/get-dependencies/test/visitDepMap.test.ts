import visitDepMap from '../src/visitDepMap';

describe('visitDepMap()', () => {
  const depMap = new Map([
    [
      "/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importAll.js",
      new Map([["b", []]])
    ], 
    [
    "/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importDefault.js",
    new Map([
      [
        "*", 
        [{
          "name": "a",
          "source": "/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importAll.js",
        }, {
          "name": "func",
          "source": "/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importAll.js",
        }],
      ],
    ]),
  ],
  [
    "/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importNamed.js",
    new Map([
      [
        "resolveFactory", 
        [{
          "name": "c",
          "source": "/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importAlias.js",
        },
        {
          "name": "func",
          "source": "/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importAlias.js",
        }],
      ]
    ])
  ]
  ]);
  test('importNamed.resolveFactory', () => {
    expect(visitDepMap(depMap, [
      {
        name: 'resolveFactory',
        source: '/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importNamed.js'
      }
    ])).toMatchSnapshot();
  });
  test('importDefault.*', () => {
    expect(visitDepMap(depMap, [
      {
        name: '*',
        source: '/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importDefault.js'
      }
    ])).toMatchSnapshot();
  });
  
  // test('circular', () => {
  //   expect(visitDepMap(
  //     new Map([
  //       [
  //         '/a',
  //         new Map([
  //           [
  //             'default',
  //             [{
  //               source: '/b',
  //               name: 'default'
  //             }]
  //           ]
  //         ])
  //       ], [
  //         '/b',
  //         new Map([
  //           [
  //             'default',
  //             [{
  //               source: '/c',
  //               name: 'default'
  //             }]
  //           ]
  //         ])
  //       ], [
  //         '/c',
  //         new Map([
  //           [
  //             'default',
  //             [{
  //               source: '/a',
  //               name: 'default'
  //             }]
  //           ]
  //         ])
  //       ]
  //     ]), 
  //     [{
  //       name: 'default',
  //       source: '/a'
  //     }]
  //   )).toMatchSnapshot();
  // });
});