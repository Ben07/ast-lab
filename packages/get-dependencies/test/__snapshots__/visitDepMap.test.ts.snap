// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`visitDepMap() circular 1`] = `
Object {
  "/a": Object {},
}
`;

exports[`visitDepMap() importDefault.* 1`] = `
Object {
  "/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importAll.js": Object {},
  "/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importDefault.js": Object {},
}
`;

exports[`visitDepMap() importNamed.resolveFactory 1`] = `
Object {
  "/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importAlias.js": Object {},
  "/Users/jennie.ji/Projects/affected/packages/get-dependencies/test/__fixtures__/imports/importNamed.js": Object {},
}
`;
